MavenPomDownloader - UncheckedIOException does not get wrapped into MavenDownloadingException

<!--
Thank you for reporting an issue with OpenRewrite!
We appreciate you taking the time to help us improve.
Please fill out the template below to help us understand and reproduce the issue.
Feel free to delete any sections that don't apply to your issue.
-->

## What version of OpenRewrite are you using?
<!--
Whenever possible please try to replicate your issue with the latest versions of OpenRewrite.
The latest major and minor versions of OpenRewrite projects are usually listed here:
https://docs.openrewrite.org/reference/latest-versions-of-every-openrewrite-module
For patch releases check the GitHub Releases page for the respective project.

We release every few weeks, so it's possible that your issue has already been fixed.

If you want to try the most recent changes that haven't been fully released yet, you can check out our snapshot releases
https://docs.openrewrite.org/reference/snapshot-instructions
-->
I am using

- rewrite @ main

## How are you running OpenRewrite?
<!--
Are you using the Maven plugin, Gradle plugin, Moderne CLI, Moderne.io or something else?
Is your project a single module or a multi-module project?

Can you share your configuration so that we can rule out any configuration issues?

Is your project public? If so, can you share a link to it?
Code snippets can also be shared privately via [our public Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA).
-->
Unit tests

## What is the smallest, simplest way to reproduce the problem?
<!--
Sometimes the logs indicate a recipe stumbled over a particular pattern of code.
If you can share a code snippet that reproduces the issue, that will help us fix it faster.
We also accept [pull requests that merely replicate an issue](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes), as a step up to a full fix.

A code snippet can be something simple like this, or similar for other languages:
-->
In my walled-garden network environment, several tests fail from `MavenPomDownloaderTest`, such as `skipsLocalInvalidArtifactsEmptyJar`, with stack traces like this:

```
org.opentest4j.AssertionFailedError: Unexpected exception type thrown, 
Expected :class org.openrewrite.maven.MavenDownloadingException
Actual   :class java.lang.RuntimeException
<Click to see difference>


	at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:67)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
	at org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3115)
	at org.openrewrite.maven.internal.MavenPomDownloaderTest.skipsLocalInvalidArtifactsEmptyJar(MavenPomDownloaderTest.java:729)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.RuntimeException: java.io.UncheckedIOException: java.net.SocketTimeoutException: Connect timed out
	at org.openrewrite.maven.internal.MavenPomDownloader.requestAsAuthenticatedOrAnonymous(MavenPomDownloader.java:810)
	at org.openrewrite.maven.internal.MavenPomDownloader.download(MavenPomDownloader.java:572)
	at org.openrewrite.maven.internal.MavenPomDownloaderTest.lambda$skipsLocalInvalidArtifactsEmptyJar$13(MavenPomDownloaderTest.java:731)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:53)
	... 6 more
Caused by: java.io.UncheckedIOException: java.net.SocketTimeoutException: Connect timed out
	at org.openrewrite.ipc.http.HttpUrlConnectionSender.send(HttpUrlConnectionSender.java:116)
	at org.openrewrite.maven.internal.MavenPomDownloader.lambda$sendRequest$1(MavenPomDownloader.java:142)
	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
	at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:74)
	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
	at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
	at org.openrewrite.maven.internal.MavenPomDownloader.sendRequest(MavenPomDownloader.java:141)
	at org.openrewrite.maven.internal.MavenPomDownloader.requestAsAuthenticatedOrAnonymous(MavenPomDownloader.java:802)
	... 9 more
Caused by: java.net.SocketTimeoutException: Connect timed out
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:551)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:602)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:633)
	at java.base/sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:304)
	at java.base/sun.net.NetworkClient.doConnect(NetworkClient.java:178)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:534)
	at java.base/sun.net.www.http.HttpClient.openServer(HttpClient.java:639)
	at java.base/sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:266)
	at java.base/sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:380)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:193)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1242)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1128)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:179)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1665)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1589)
	at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:529)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:308)
	at org.openrewrite.ipc.http.HttpUrlConnectionSender.send(HttpUrlConnectionSender.java:97)
	... 18 more

```

## What did you expect to see?
<!-- A code snippet, or a description of the behavior you expected helps us write a test to ensure the issue is fixed. -->
Happy test, or, perhaps an easy-to-read failure from being unable to connect to the default remote Maven repo in my environment.

## What did you see instead?
<!-- A code snippet, or a description of the behavior you saw instead of the above expected result. -->
The resultant exception is not wrapped into `MavenDownloadingException` and therefore does not give useful detail about the exact request/artifact that led to an error.

So when this happens for real recipe runs, that lack of detail is really inconvenient.

## What is the full stack trace of any errors you encountered?
<!-- When errors occur, please include the output of `--stacktrace` for Gradle or `--debug` for Maven. -->
Seen above

## Are you interested in [contributing a fix to OpenRewrite](https://github.com/openrewrite/.github/blob/main/CONTRIBUTING.md#contributing-fixes)?
<!-- Indicate if this is something you would like to work on, and how we can best support you in doing so. -->
Yes, but, it's a little tricky because I know I'm in a non-standard network. I'd ideally like to define a unit test which simulates the "connect timeout" behavior, but, hard for me to know for sure that any given test will achieve that same behavior both in my network and in a "standard" network. I might have to spam Actions a little more than usual to validate that.

--- a/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
+++ b/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
@@ -771,8 +771,6 @@ public class MavenPomDownloader {
                                             repository.getSnapshots(),
                                             repository.getUsername(),
                                             repository.getPassword());
-                                } else if (!e.isClientSideException()) {
-                                    return originalRepository;
                                 }
                             } catch (Throwable e) {
                                 // ok to fall through here and cache a null

--- a/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
+++ b/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
@@ -151,8 +151,7 @@ public class MavenPomDownloader {
             throw failsafeException.getCause() == null ? failsafeException : failsafeException.getCause();
         } catch (Throwable e) {
             throw new HttpSenderResponseException(e, null, "");
-        }
-        finally {
+        } finally {
             this.ctx.recordResolutionTime(Duration.ofNanos(System.nanoTime() - start));
         }
     }

--- a/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
+++ b/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
@@ -149,7 +149,10 @@ public class MavenPomDownloader {
             });
         } catch (FailsafeException failsafeException) {
             throw failsafeException.getCause() == null ? failsafeException : failsafeException.getCause();
-        } finally {
+        } catch (Throwable e) {
+            throw new HttpSenderResponseException(e, null, "");
+        }
+        finally {
             this.ctx.recordResolutionTime(Duration.ofNanos(System.nanoTime() - start));
         }
     }

--- a/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
+++ b/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
@@ -284,7 +284,7 @@ public class MavenPomDownloader {
                                     .increment();
                             result = Optional.of(derivedMeta);
                         }
-                    } catch (HttpSenderResponseException | MavenDownloadingException e) {
+                    } catch (HttpSenderResponseException | MavenDownloadingException | IOException e) {
                         repositoryResponses.put(repo, e.getMessage());
                     }
                 }
@@ -330,7 +330,7 @@ public class MavenPomDownloader {
      * @return Metadata or null if the metadata cannot be derived.
      */
     @Nullable
-    private MavenMetadata deriveMetadata(GroupArtifactVersion gav, MavenRepository repo) throws HttpSenderResponseException, MavenDownloadingException {
+    private MavenMetadata deriveMetadata(GroupArtifactVersion gav, MavenRepository repo) throws HttpSenderResponseException, IOException, MavenDownloadingException {
         if ((repo.getDeriveMetadataIfMissing() != null && !repo.getDeriveMetadataIfMissing()) || gav.getVersion() != null) {
             // Do not derive metadata if we cannot navigate/browse the artifacts.
             // Do not derive metadata if a specific version has been defined.
@@ -592,6 +592,8 @@ public class MavenPomDownloader {
                             //If the exception is a common, client-side exception, cache an empty result.
                             mavenCache.putPom(resolvedGav, null);
                         }
+                    } catch (IOException e) {
+                        repositoryResponses.put(repo, e.getMessage());
                     }
                 }
             } else if (result.isPresent()) {
@@ -797,7 +799,7 @@ public class MavenPomDownloader {
     /**
      * Replicates Apache Maven's behavior to attempt anonymous download if repository credentials prove invalid
      */
-    private byte[] requestAsAuthenticatedOrAnonymous(MavenRepository repo, String uriString) throws HttpSenderResponseException {
+    private byte[] requestAsAuthenticatedOrAnonymous(MavenRepository repo, String uriString) throws HttpSenderResponseException, IOException {
         try {
             return sendRequest(applyAuthenticationToRequest(repo, httpSender.get(uriString)).build());
         } catch (HttpSenderResponseException e) {
@@ -806,12 +808,14 @@ public class MavenPomDownloader {
             } else {
                 throw e;
             }
+        } catch (IOException e) {
+            throw e;
         } catch (Throwable t) {
             throw new RuntimeException(t);  // unreachable
         }
     }
 
-    private byte[] retryRequestAnonymously(String uriString, HttpSenderResponseException originalException) throws HttpSenderResponseException {
+    private byte[] retryRequestAnonymously(String uriString, HttpSenderResponseException originalException) throws HttpSenderResponseException, IOException {
         try {
             return sendRequest(httpSender.get(uriString).build());
         } catch (HttpSenderResponseException retryException) {
@@ -820,6 +824,8 @@ public class MavenPomDownloader {
             } else {
                 throw retryException;
             }
+        } catch (IOException e) {
+            throw e;
         } catch (Throwable t) {
             throw new RuntimeException(t);  // unreachable
         }

--- a/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
+++ b/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
@@ -149,8 +149,8 @@ public class MavenPomDownloader {
             });
         } catch (FailsafeException failsafeException) {
             throw failsafeException.getCause() == null ? failsafeException : failsafeException.getCause();
-        } catch (Throwable e) {
-            throw new HttpSenderResponseException(e, null, "");
+        } catch (UncheckedIOException e) {
+            throw e.getCause();
         } finally {
             this.ctx.recordResolutionTime(Duration.ofNanos(System.nanoTime() - start));
         }

--- a/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
+++ b/rewrite-maven/src/main/java/org/openrewrite/maven/internal/MavenPomDownloader.java
@@ -135,7 +135,7 @@ public class MavenPomDownloader {
         this.mavenCache = this.ctx.getPomCache();
     }
 
-    byte[] sendRequest(HttpSender.Request request) throws Throwable {
+    byte[] sendRequest(HttpSender.Request request) throws IOException, HttpSenderResponseException {
         long start = System.nanoTime();
         try {
             return Failsafe.with(retryPolicy).get(() -> {
@@ -148,7 +148,10 @@ public class MavenPomDownloader {
                 }
             });
         } catch (FailsafeException failsafeException) {
-            throw failsafeException.getCause() == null ? failsafeException : failsafeException.getCause();
+            if (failsafeException.getCause() instanceof HttpSenderResponseException) {
+                throw (HttpSenderResponseException) failsafeException.getCause();
+            }
+            throw failsafeException;
         } catch (UncheckedIOException e) {
             throw e.getCause();
         } finally {
@@ -808,10 +811,6 @@ public class MavenPomDownloader {
             } else {
                 throw e;
             }
-        } catch (IOException e) {
-            throw e;
-        } catch (Throwable t) {
-            throw new RuntimeException(t);  // unreachable
         }
     }
 
@@ -824,10 +823,6 @@ public class MavenPomDownloader {
             } else {
                 throw retryException;
             }
-        } catch (IOException e) {
-            throw e;
-        } catch (Throwable t) {
-            throw new RuntimeException(t);  // unreachable
         }
     }
 

--- a/rewrite-maven/src/test/java/org/openrewrite/maven/internal/MavenPomDownloaderTest.java
+++ b/rewrite-maven/src/test/java/org/openrewrite/maven/internal/MavenPomDownloaderTest.java
@@ -36,6 +36,7 @@ import org.openrewrite.maven.tree.*;
 
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
+import java.net.UnknownHostException;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -187,7 +188,7 @@ class MavenPomDownloaderTest {
             // not expected to succeed
         }
         assertThat(attemptedUris).isNotEmpty();
-        assertThat(attemptedUris.get(httpUrl)).hasMessageContaining("java.net.UnknownHostException");
+        assertThat(attemptedUris.get(httpUrl)).isInstanceOf(UnknownHostException.class);
         assertThat(discoveredRepositories).isEmpty();
     }
 

--- a/rewrite-maven/src/test/java/org/openrewrite/maven/internal/MavenPomDownloaderTest.java
+++ b/rewrite-maven/src/test/java/org/openrewrite/maven/internal/MavenPomDownloaderTest.java
@@ -48,7 +48,6 @@ import static java.util.Collections.emptyList;
 import static java.util.Collections.emptyMap;
 import static java.util.Collections.singletonList;
 import static java.util.Collections.singletonMap;
-import static java.util.Objects.requireNonNull;
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
@@ -86,7 +85,7 @@ class MavenPomDownloaderTest {
           null, "true", false, null, null, null);
         MavenRepository repo = new MavenPomDownloader(ctx).normalizeRepository(ossSonatype,
           MavenExecutionContextView.view(ctx), null);
-        assertThat(requireNonNull(repo).getUri()).isEqualTo(ossSonatype.getUri());
+        assertThat(repo).isNotNull().extracting((MavenRepository::getUri)).isEqualTo(ossSonatype.getUri());
     }
 
     @Issue("https://github.com/openrewrite/rewrite/issues/3908")
@@ -740,6 +739,20 @@ class MavenPomDownloaderTest {
         var result = downloader.download(gav, null, null, List.of());
         assertThat(result.getRepository()).isNotNull();
         assertThat(result.getRepository().getUri()).startsWith(tempDir.toUri().toString());
+    }
+
+    @Issue("https://github.com/openrewrite/rewrite/issues/4080")
+    @Test
+    void connectTimeout() {
+        var downloader = new MavenPomDownloader(ctx);
+        var gav = new GroupArtifactVersion("org.openrewrite", "rewrite-core", "7.0.0");
+        var repos = singletonList(MavenRepository.builder()
+          .id("non-routable").uri("http://10.0.0.0/maven").knownToExist(true).build());
+
+        assertThatThrownBy(() -> downloader.download(gav, null, null, repos))
+          .isInstanceOf(MavenDownloadingException.class)
+          .hasMessageContaining("rewrite-core")
+          .hasMessageContaining("10.0.0.0");
     }
 
     private static GroupArtifactVersion createArtifact(Path repository) throws IOException {

